clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -3056,11 +3071,16 @@ class SYCLKernelNameTypeVisitor
               << /* kernel name is not globally-visible */ 0
               << QualType(Tag->getTypeForDecl(), 0);
           IsInvalid = true;
-        } else {
+          return;
+        }
+        if (isa<CXXMethodDecl>(DeclCtx) && !Tag->isCompleteDefinition()) {
```

Comment:
This and line 3065 seems to change the logic quite a bit here.  Can you explain what is happening here?

Reply:
The if block at L3076 is to handle the cases like:
```
        int main()
        h.single_task<class fake_kernel2>([]() {
              auto l = [](auto f) { f(); };
        });
```
The order of DeclContexts for fake_kernel2 is
**clang::Decl::CXXMethod**
clang::Decl::CXXRecord
clang::Decl::Function
clang::Decl::TranslationUnit

So I just check if the DeclCtx is clang::Decl::CXXMethod and "class fake_kernel2" does not have complete definition.